File: apps/docs/src/prework/examples/object-oriented.ts
import type { Reason, Request } from "./procedural";

interface Rule {
  check(request: Request): readonly Reason[];
}

export class EquipmentRule implements Rule {
  check(request: Request): readonly Reason[] {
    return request.equipmentChecked ? [] : ["EquipmentNotChecked"];
  }
}

export class FlareRule implements Rule {
  check(request: Request): readonly Reason[] {
    return request.flareAlert ? ["FlareAlertActive"] : [];
  }
}

export class ApprovalPolicy {
  constructor(private readonly rules: readonly Rule[]) {}

  check(request: Request): readonly Reason[] {
    return this.rules.flatMap((rule) => rule.check(request));
  }
}

export const policy = new ApprovalPolicy([
  new EquipmentRule(),
  new FlareRule(),
]);

File: apps/docs/src/prework/examples/rust/object_oriented.rs
use crate::procedural::{Reason, Request};

pub trait Rule {
    fn check(&self, request: &Request) -> Vec<Reason>;
}

pub struct EquipmentRule;
impl Rule for EquipmentRule {
    fn check(&self, request: &Request) -> Vec<Reason> {
        if request.equipment_checked {
            vec![]
        } else {
            vec![Reason::EquipmentNotChecked]
        }
    }
}

pub struct FlareRule;
impl Rule for FlareRule {
    fn check(&self, request: &Request) -> Vec<Reason> {
        if request.flare_alert {
            vec![Reason::FlareAlertActive]
        } else {
            vec![]
        }
    }
}

pub struct ApprovalPolicy {
    pub rules: Vec<Box<dyn Rule>>,
}

impl ApprovalPolicy {
    pub fn check(&self, request: &Request) -> Vec<Reason> {
        self.rules
            .iter()
            .flat_map(|rule| rule.check(request))
            .collect()
    }
}

pub fn policy() -> ApprovalPolicy {
    ApprovalPolicy {
        rules: vec![Box::new(EquipmentRule), Box::new(FlareRule)],
    }
}
